File: go/arrays/merge_solution.go
package arrays

func MergeSolution(arr []int, m int) {
	n := len(arr)
	buf := make([]int, n)
	for i, j, k := 0, m, 0; k < n; k++ {
		if j == n || (i < m && arr[i] <= arr[j]) {
			buf[k] = arr[i]
			i++
		} else {
			buf[k] = arr[j]
			j++
		}
	}
	copy(arr, buf)
}

File: java/src/arrays/MergeSolution.java
package arrays;

public class MergeSolution {
	public static void merge(int[] arr, int l, int m, int r) {
		int n = r - l;
		int[] buf = new int[n];
		for (int i = l, j = m, k = 0; k < n; k++) {
			if (j == r || (i < m && arr[i] <= arr[j])) {
				buf[k] = arr[i];
				i++;
			} else {
				buf[k] = arr[j];
				j++;
			}
		}
		System.arraycopy(buf, 0, arr, l, n);
	}
}
